### PR TITLE
chore: refactor

### PR DESCRIPTION
**Pull request by simlu**

### CODE REVIEW

File: test/util/iterator.spec.js [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+          index = 0;
+        }
+        if (!(index < CHARS.length) || Math.random() * MAX_DEPTH < depth) {
+          index += 1;
+          return CHARS[(index - 1) % CHARS.length];
+        }
+        const result = Math.random() > 0.5 ? new Set() : [];
+        for (let idx = 0, len = Math.ceil(Math.random() * MAX_WIDTH); idx < len; idx += 1) {
+          const entry = treeGen(depth + 1);
+          result[Array.isArray(result) ? 'push' : 'add'](entry);
```
Comment (simlu) [inline test/util/iterator.spec.js:27]: Also test empty set / list (?)

File: src/util/compiler.js [hunk 2]
```diff
@@ -55,41 +56,67 @@ module.exports.matchedBy = (searches) => extractNeedles(searches.filter((e) => i
...
-    Object.assign(tower, { [path[0]]: {} });
-    if (String(path[0]) === '**') {
-      markRecursive(tower[path[0]]);
+    cb(type, p, excluded);
+  });
+};
+
+const applyNeedle = (tower, needle, strict) => {
+  const tree = [parser.parse(needle)];
+  const stack = [[tower]];
```
Comment (simlu) [inline src/util/compiler.js:81]: Abstract more into the iterate and only expose onFin on onAdd cb functions

File: src/util/iterator.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
+const iterate = (tree, cb) => {
+  const stack = [tree];
+  const parent = [null];
+  const count = [];
+  const depth = [];
+  const path = [];
+  let idx = 0;
+  let inc = true;
+
+  while (idx >= 0) {
```
Comment (simlu) [inline src/util/iterator.js:10]: ! == - 1

File: src/util/compiler.js [hunk 4]
```diff
@@ -55,41 +56,68 @@ module.exports.matchedBy = (searches) => extractNeedles(searches.filter((e) => i
...
+      }
+      stack.pop();
+    } else if (type === 'ADD') {
+      const isExcluded = p.isExcluded();
+      if (excluded && isExcluded) {
+        throw new Error(`Redundant Exclusion: "${needle}"`);
+      }
+      if (isExcluded) {
+        excluded = true;
+      }
```
Comment (simlu) [inline src/util/compiler.js:77]:
```suggestion
      if (p.isExcluded()) {
        if (excluded) {
          throw new Error(`Redundant Exclusion: "${needle}"`);
        }
        excluded = true;
      }
```
